src/main/java/com/cosmian/jna/findex/Ffi.java

```diff
@@ -0,0 +1,145 @@
+    }
+
+    /**
+     * Set the last error on the native lib
+     *
+     * @param error_msg the last error to set on the native lib
+     * @throws FfiException n case of native library error
+     */
+    public static void set_error(String error_msg) throws FfiException {
+        unwrap(Ffi.INSTANCE.set_error(error_msg));
+    }
+
+    public static void upsert(MasterKeys masterKeys, HashMap<String, String[]> dbUidsAndWords,
+        FetchEntryCallback fetchEntry, UpsertEntryCallback upsertEntry, UpsertChainCallback upsertChain)
+        throws FfiException, CosmianException {
+
+        // For the JSON strings
+        ObjectMapper mapper = new ObjectMapper();
+
+        // Findex master keys
+        String masterKeysJson;
+        try {
+            masterKeysJson = mapper.writeValueAsString(masterKeys);
+        } catch (JsonProcessingException e) {
+            throw new FfiException("Invalid master keys");
+        }
+
+        // Findex db UIDS and words
+        String dbUidsAndWordsJson;
+        try {
+            dbUidsAndWordsJson = mapper.writeValueAsString(dbUidsAndWords);
+        } catch (JsonProcessingException e) {
+            throw new FfiException("Invalid db uids and words");
+        }
+
+        // Indexes creation + insertion/update
+        unwrap(Ffi.INSTANCE.h_upsert(masterKeysJson, dbUidsAndWordsJson, fetchEntry, upsertEntry, upsertChain));
+    }
+
+    public static String[] search(MasterKeys masterKeys, String[] words, int loopIterationLimit,
+        FetchEntryCallback fetchEntry, FetchChainCallback fetchChain) throws FfiException, CosmianException {
+        //
+        // Prepare outputs
+        //
+        byte[] dbUidsBuffer = new byte[8192];
+        IntByReference dbUidsBufferSize = new IntByReference(dbUidsBuffer.length);
+
+        // For the JSON strings
+        ObjectMapper mapper = new ObjectMapper();
+
+        // Findex master keys
+        String masterKeysJson;
+        try {
+            masterKeysJson = mapper.writeValueAsString(masterKeys);
+        } catch (JsonProcessingException e) {
+            throw new FfiException("Invalid master keys");
+        }
+
+        // Findex words
+        String wordsJson;
+        try {
+            wordsJson = mapper.writeValueAsString(words);
+        } catch (JsonProcessingException e) {
+            throw new FfiException("Invalid words");
+        }
+
+        String[] dbUids = null;
+
+        // Indexes creation + insertion/update
+        unwrap(Ffi.INSTANCE.h_search(dbUidsBuffer, dbUidsBufferSize, masterKeysJson, wordsJson, loopIterationLimit,
+            fetchEntry, fetchChain));
+
+        byte[] dbUidsBytes = Arrays.copyOfRange(dbUidsBuffer, 0, dbUidsBufferSize.getValue());
+        try {
+            dbUids = mapper.readValue(dbUidsBytes, String[].class);
+        } catch (IOException e) {
+            throw new FfiException("DB UIDs deserialization failed");
+        }
+        Arrays.sort(dbUids);
```

Comment:
Sorting makes sense if you run conjunction afterward: the algorithm is O(n) instead of O(n^ 2). Since sorting is O(nlog(n)) the sorted version is asymptotically O(nlog(n)) which is much better than O(n^2)

Reply:
it was for test purposes indeed